internal/agg/aggregator.go

```diff
@@ -0,0 +1,180 @@
+package agg
+
+import (
+	"encoding/json"
+	"errors"
+	"log"
+	"slices"
+	"time"
+
+	"github.com/PDOK/azure-storage-usage-exporter/internal/du"
+
+	_ "github.com/marcboeker/go-duckdb" // duckdb sql driver
+
+	"golang.org/x/exp/maps"
+)
+
+const (
+	Deleted = "deleted"
+)
+
+type Labels = map[string]string
+
+type AggregationRule struct {
+	// The named groups are used as labels
+	Pattern ReGroup `yaml:"pattern"`
+	// A label not found as named group is looked up in this
+	StaticLabels map[string]string `yaml:"labels"`
+}
+
+type AggregationGroup struct {
+	Labels  Labels
+	Deleted bool
+}
+
+type AggregationResult struct {
+	AggregationGroup AggregationGroup
+	StorageUsage     du.StorageUsage
+}
+
+type Aggregator struct {
+	duReader           du.Reader
+	labelsWithDefaults Labels
+	rules              []AggregationRule
+}
+
+func NewAggregator(duReader du.Reader, labels Labels, rules []AggregationRule) (*Aggregator, error) {
+	if _, exists := labels[Deleted]; exists {
+		return nil, errors.New("cannot use deleted as a label")
+	}
+	return &Aggregator{
+		duReader:           duReader,
+		labelsWithDefaults: labels,
+		rules:              rules,
+	}, nil
+}
+
+func (a *Aggregator) GetLabelNames() []string {
+	keys := maps.Keys(a.labelsWithDefaults)
+	keys = append(keys, Deleted)
+	return keys
+}
+
+func (a *Aggregator) Aggregate(previousRunDate time.Time) (aggregationResults []AggregationResult, runDate time.Time, err error) {
+	log.Print("starting aggregation")
+	runDate, rowsCh, errCh, err := a.duReader.Read(previousRunDate)
+	if err != nil {
+		return nil, runDate, err
+	}
+	if !runDate.After(previousRunDate) {
+		return nil, runDate, nil
+	}
+
+	intermediateResults := make(map[string]du.StorageUsage)
+	i := 0
+	for rowsCh != nil && errCh != nil {
+		select {
+		case err, ok := <-errCh:
+			if !ok {
+				errCh = nil
+				continue
+			}
+			if err != nil {
+				return nil, runDate, err
+			}
+		case row, ok := <-rowsCh:
+			if !ok {
+				rowsCh = nil
+				continue
+			}
+			aggregationGroup := a.applyRulesToAggregate(row)
+			intermediateResults[marshalAggregationGroup(aggregationGroup)] += row.Bytes
+			if i%10000 == 0 {
+				log.Printf("%d du rows processed so far", i)
+			}
+			i++
+		}
+	}
+	log.Printf("done aggregating blob inventory, %d du rows processed", i)
```

Comment:
idem: `du` moest even nadenken wat du betekende (dacht eerst aan **du**ckdb), maar het is disk usage. Zou het in de logging uitschrijven als disk usage